File: sheets_handler.py
"""
Sheets Handler - Manages all Google Sheets operations
"""
import gspread
from datetime import datetime


def get_sheets_client(credentials):
    """Create and return gspread client"""
    return gspread.authorize(credentials)


def open_spreadsheet(gc, spreadsheet_id):
    """Open spreadsheet by ID"""
    return gc.open_by_key(spreadsheet_id)


def get_worksheet(sheet, worksheet_name):
    """Get a specific worksheet"""
    return sheet.worksheet(worksheet_name)


def get_all_tickets(worksheet):
    """
    Get all existing tickets from the sheet
    Returns: dict mapping thread_id to row_number
    """
    rows = worksheet.get_all_values()
    thread_map = {}
    
    for i, row in enumerate(rows[1:], start=2):  # Skip header, start from row 2
        if len(row) > 1 and row[1]:  # Check if thread_id exists
            thread_map[row[1]] = i
    
    return thread_map


def get_next_ticket_id(sheet):
    """
    Generate next ticket ID
    Reads from Ticket_Config sheet and increments counter
    """
    config_sheet = sheet.worksheet("Ticket_Config")
    last_ticket = int(config_sheet.acell("B1").value or 0)
    next_ticket = last_ticket + 1
    
    # Update counter
    config_sheet.update(range_name="B1", values=[[next_ticket]])
    
    return f"TCK-{next_ticket:06d}"


def create_ticket_row(ticket_id, thread_id, from_email, subject, status, new_sender=False):
    """
    Create a ticket row for the spreadsheet
    Returns: list of values for the row
    """
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    link = f'=HYPERLINK("https://mail.google.com/mail/u/0/#inbox/{thread_id}","Open Mail")'
    
    return [
        ticket_id,
        thread_id,
        timestamp,
        from_email,
        subject,
        status,
        "Yes" if new_sender else "No",  # Dynamic new sender detection
        link,
    ]


def add_new_ticket(worksheet, row_data):
    """Add a new ticket row to the sheet"""
    worksheet.append_row(row_data, value_input_option="USER_ENTERED")


def update_existing_ticket(worksheet, row_number, row_data):
    """Update an existing ticket row"""
    range_name = f"A{row_number}:H{row_number}"
    worksheet.update(range_name, [row_data], value_input_option="USER_ENTERED")


def get_ticket_data(worksheet, row_number):
    """Get ticket data from a specific row"""
    return worksheet.row_values(row_number)


# ================= SYNC STATE MANAGEMENT =================
def initialize_state_sheets(sheet):
    """
    Initialize Sync_State and Thread_State sheets if they don't exist
    """
    # Initialize Sync_State
    try:
        sheet.worksheet("Sync_State")
        print("📊 Sync_State sheet already exists")
    except:
        sync_sheet = sheet.add_worksheet(title="Sync_State", rows=10, cols=2)
        sync_sheet.update("A1", [["Last Sync"]])
        print("✅ Created Sync_State sheet")
    
    # Initialize Thread_State
    try:
        sheet.worksheet("Thread_State")
        print("📊 Thread_State sheet already exists")
    except:
        thread_sheet = sheet.add_worksheet(title="Thread_State", rows=1000, cols=2)
        thread_sheet.update("A1", [["Thread ID", "Last Processed Timestamp"]])
        print("✅ Created Thread_State sheet")


def get_last_sync_timestamp(sheet):
    """
    Get last sync timestamp from Sync_State sheet
    Returns: timestamp (int) or None
    """
    try:
        sync_sheet = sheet.worksheet("Sync_State")
        value = sync_sheet.acell("B1").value
        return int(value) if value else None
    except:
        # Sheet doesn't exist or no value
        return None


def save_last_sync_timestamp(sheet, timestamp):
    """
    Save last sync timestamp to Sync_State sheet
    """
    try:
        sync_sheet = sheet.worksheet("Sync_State")
    except:
        # Create sheet if it doesn't exist
        sync_sheet = sheet.add_worksheet(title="Sync_State", rows=10, cols=2)
        # Set header
        sync_sheet.update("A1", [["Last Sync"]])
    
    # Update timestamp
    sync_sheet.update("B1", [[timestamp]])


# ================= THREAD STATE MANAGEMENT =================
def load_thread_state_from_sheet(sheet):
    """
    Load thread processing state from Thread_State sheet
    Returns: dict mapping thread_id to timestamp
    """
    state = {}
    try:
        thread_sheet = sheet.worksheet("Thread_State")
        rows = thread_sheet.get_all_values()
        
        # Skip header row
        for row in rows[1:]:
            if len(row) >= 2 and row[0]:
                state[row[0]] = int(row[1])
    except:
        # Sheet doesn't exist yet
        pass
    
    return state


def save_thread_state_to_sheet(sheet, state):
    """
    Save thread processing state to Thread_State sheet
    """
    try:
        thread_sheet = sheet.worksheet("Thread_State")
    except:
        # Create sheet if it doesn't exist
        thread_sheet = sheet.add_worksheet(title="Thread_State", rows=1000, cols=2)
    
    # Convert state dict to list of lists
    data = [["Thread ID", "Last Processed Timestamp"]]
    for tid, ts in state.items():
        data.append([tid, ts])
    
    # Clear and update
    thread_sheet.clear()
    thread_sheet.update("A1", data)

File: SheetsHandler.php
<?php
/**
 * SheetsHandler - Manages all Google Sheets operations
 */

declare(strict_types=1);

namespace GmailTicket;

use Google\Client;
use Google\Service\Sheets;
use Google\Service\Sheets\ValueRange;
use Google\Service\Sheets\Spreadsheet;
use Google\Service\Sheets\Sheet;
use Google\Service\Sheets\Request;
use Google\Service\Sheets\BatchUpdateSpreadsheetRequest;
use Google\Service\Sheets\AddSheetRequest;
use Google\Service\Sheets\SheetProperties;

class SheetsHandler
{
    private Sheets $service;

    public function __construct(Client $client)
    {
        $this->service = new Sheets($client);
    }

    // ──────────────────────────────────────────────────────
    //  Spreadsheet / worksheet helpers
    // ──────────────────────────────────────────────────────

    /**
     * Read all values from a worksheet range.
     *
     * @return array<int, array<int, string>>
     */
    public function getValues(string $spreadsheetId, string $range): array
    {
        $response = $this->service->spreadsheets_values->get($spreadsheetId, $range);
        return $response->getValues() ?? [];
    }

    /**
     * Write a 2-D array to a range (USER_ENTERED so formulas are evaluated).
     *
     * @param array<int, array<int, mixed>> $values
     */
    public function setValues(string $spreadsheetId, string $range, array $values): void
    {
        $body = new ValueRange(['values' => $values]);
        $this->service->spreadsheets_values->update(
            $spreadsheetId,
            $range,
            $body,
            ['valueInputOption' => 'USER_ENTERED']
        );
    }

    /**
     * Append a single row (USER_ENTERED).
     *
     * @param array<int, mixed> $row
     */
    public function appendRow(string $spreadsheetId, string $range, array $row): void
    {
        $body = new ValueRange(['values' => [$row]]);
        $this->service->spreadsheets_values->append(
            $spreadsheetId,
            $range,
            $body,
            ['valueInputOption' => 'USER_ENTERED']
        );
    }

    /**
     * Read a single cell value.
     */
    public function getCell(string $spreadsheetId, string $cell): ?string
    {
        $values = $this->getValues($spreadsheetId, $cell);
        return $values[0][0] ?? null;
    }

    /**
     * Write a single cell value.
     */
    public function setCell(string $spreadsheetId, string $cell, mixed $value): void
    {
        $this->setValues($spreadsheetId, $cell, [[$value]]);
    }

    /**
     * Clear all values in a range.
     */
    public function clearRange(string $spreadsheetId, string $range): void
    {
        $this->service->spreadsheets_values->clear($spreadsheetId, $range, new \Google\Service\Sheets\ClearValuesRequest());
    }

    /**
     * Ensure a worksheet exists; create it if missing.
     * Returns the sheet title (same as $name).
     */
    public function ensureWorksheet(string $spreadsheetId, string $name, int $rows = 1000, int $cols = 2): string
    {
        // Check existing sheets
        $meta = $this->service->spreadsheets->get($spreadsheetId);
        foreach ($meta->getSheets() as $sheet) {
            if ($sheet->getProperties()->getTitle() === $name) {
                return $name; // Already exists
            }
        }

        // Create new sheet
        $addSheet = new AddSheetRequest([
            'properties' => new SheetProperties([
                'title'     => $name,
                'gridProperties' => [
                    'rowCount'    => $rows,
                    'columnCount' => $cols,
                ],
            ]),
        ]);
        $batchReq = new BatchUpdateSpreadsheetRequest([
            'requests' => [new Request(['addSheet' => $addSheet])],
        ]);
        $this->service->spreadsheets->batchUpdate($spreadsheetId, $batchReq);

        return $name;
    }

    // ──────────────────────────────────────────────────────
    //  Ticket map
    // ──────────────────────────────────────────────────────

    /**
     * Return a map of thread_id => row_number (1-indexed, skipping header).
     *
     * @return array<string, int>
     */
    public function getAllTickets(string $spreadsheetId, string $worksheet = 'Email log'): array
    {
        $rows = $this->getValues($spreadsheetId, "{$worksheet}!A:B");
        $map  = [];

        foreach ($rows as $idx => $row) {
            if ($idx === 0) {
                continue; // Skip header
            }
            $rowNum  = $idx + 1;       // 1-indexed
            $threadId = $row[1] ?? ''; // Column B
            if ($threadId !== '') {
                $map[$threadId] = $rowNum;
            }
        }

        return $map;
    }

    /**
     * Get a specific row's values.
     *
     * @return array<int, string>
     */
    public function getRow(string $spreadsheetId, string $worksheet, int $rowNumber): array
    {
        $range = "{$worksheet}!A{$rowNumber}:H{$rowNumber}";
        $values = $this->getValues($spreadsheetId, $range);
        return $values[0] ?? [];
    }

    // ──────────────────────────────────────────────────────
    //  Ticket ID counter
    // ──────────────────────────────────────────────────────

    /**
     * Read counter from Ticket_Config!B1, increment, save, and return "TCK-XXXXXX"
     */
    public function getNextTicketId(string $spreadsheetId): string
    {
        $raw  = $this->getCell($spreadsheetId, 'Ticket_Config!B1');
        $next = ((int) $raw) + 1;
        $this->setCell($spreadsheetId, 'Ticket_Config!B1', $next);
        return sprintf('TCK-%06d', $next);
    }

    // ──────────────────────────────────────────────────────
    //  Row builders
    // ──────────────────────────────────────────────────────

    /**
     * Build a ticket row ready for insertion / update.
     *
     * @return array<int, mixed>
     */
    public function createTicketRow(
        string $ticketId,
        string $threadId,
        string $fromEmail,
        string $subject,
        string $status,
        bool   $newSender = false
    ): array {
        $timestamp = date('Y-m-d H:i:s');
        $link      = sprintf(
            '=HYPERLINK("https://mail.google.com/mail/u/0/#inbox/%s","Open Mail")',
            $threadId
        );

        return [
            $ticketId,
            $threadId,
            $timestamp,
            $fromEmail,
            $subject,
            $status,
            $newSender ? 'Yes' : 'No',
            $link,
        ];
    }

    /**
     * Append a new ticket row to the worksheet.
     *
     * @param array<int, mixed> $rowData
     */
    public function addNewTicket(string $spreadsheetId, string $worksheet, array $rowData): void
    {
        $this->appendRow($spreadsheetId, "{$worksheet}!A:H", $rowData);
    }

    /**
     * Overwrite an existing ticket row.
     *
     * @param array<int, mixed> $rowData
     */
    public function updateExistingTicket(string $spreadsheetId, string $worksheet, int $rowNumber, array $rowData): void
    {
        $range = "{$worksheet}!A{$rowNumber}:H{$rowNumber}";
        $this->setValues($spreadsheetId, $range, [$rowData]);
    }

    // ──────────────────────────────────────────────────────
    //  Sync timestamp (Sync_State sheet)
    // ──────────────────────────────────────────────────────

    public function initializeStateSheets(string $spreadsheetId): void
    {
        // Sync_State
        $this->ensureWorksheet($spreadsheetId, 'Sync_State', 10, 2);
        $this->setValues($spreadsheetId, 'Sync_State!A1', [['Last Sync']]);

        // Thread_State
        $this->ensureWorksheet($spreadsheetId, 'Thread_State', 1000, 2);
        $this->setValues($spreadsheetId, 'Thread_State!A1', [['Thread ID', 'Last Processed Timestamp']]);
    }

    public function getLastSyncTimestamp(string $spreadsheetId): ?int
    {
        $value = $this->getCell($spreadsheetId, 'Sync_State!B1');
        return ($value !== null && $value !== '') ? (int) $value : null;
    }

    public function saveLastSyncTimestamp(string $spreadsheetId, int $timestamp): void
    {
        $this->ensureWorksheet($spreadsheetId, 'Sync_State', 10, 2);
        $this->setCell($spreadsheetId, 'Sync_State!B1', $timestamp);
    }

    // ──────────────────────────────────────────────────────
    //  Thread state (Thread_State sheet)
    // ──────────────────────────────────────────────────────

    /**
     * Load thread processing state from the Thread_State worksheet.
     *
     * @return array<string, int>   thread_id => timestamp
     */
    public function loadThreadStateFromSheet(string $spreadsheetId): array
    {
        $state = [];
        try {
            $rows = $this->getValues($spreadsheetId, 'Thread_State!A:B');
            foreach ($rows as $idx => $row) {
                if ($idx === 0 || count($row) < 2 || $row[0] === '') {
                    continue;
                }
                $state[$row[0]] = (int) $row[1];
            }
        } catch (\Exception) {
            // Sheet might not exist yet — that's fine
        }
        return $state;
    }

    /**
     * Persist thread state to the Thread_State worksheet.
     *
     * @param array<string, int> $state
     */
    public function saveThreadStateToSheet(string $spreadsheetId, array $state): void
    {
        $this->ensureWorksheet($spreadsheetId, 'Thread_State', 1000, 2);

        $data = [['Thread ID', 'Last Processed Timestamp']];
        foreach ($state as $tid => $ts) {
            $data[] = [$tid, $ts];
        }

        $this->clearRange($spreadsheetId, 'Thread_State!A:B');
        $this->setValues($spreadsheetId, 'Thread_State!A1', $data);
    }

   
    /**
     * @return string[]
     */


    // ──────────────────────────────────────────────────────
    //  Known senders
    // ──────────────────────────────────────────────────────

    /**
     * Read column D (index 3) from the Email log and return all unique emails.
     *
     * @return array<string, true>   email => true (set-like)
     */
    public function loadKnownSenders(string $spreadsheetId, string $worksheet = 'Email log'): array
    {
        $rows    = $this->getValues($spreadsheetId, "{$worksheet}!A:H");
        $senders = [];
        foreach ($rows as $idx => $row) {
            if ($idx === 0) {
                continue; // Skip header
            }
            $email = strtolower($row[3] ?? '');
            if ($email !== '') {
                $senders[$email] = true;
            }
        }
        return $senders;
    }

    // ──────────────────────────────────────────────────────
    //  All rows (for stale-ticket check)
    // ──────────────────────────────────────────────────────

    /**
     * @return array<int, array<int, string>>
     */
    public function getAllRows(string $spreadsheetId, string $worksheet = 'Email log'): array
    {
        return $this->getValues($spreadsheetId, "{$worksheet}!A:H");
    }

    /**
     * Update a single row in-place (used by auto-close).
     *
     * @param array<int, string> $row
     */
    public function updateRow(string $spreadsheetId, string $worksheet, int $rowNumber, array $row): void
    {
        $range = "{$worksheet}!A{$rowNumber}:H{$rowNumber}";
        $this->setValues($spreadsheetId, $range, [$row]);
    }

    /**
     * Delete a single row by index (used by auto-close with "delete" action).
     * NOTE: Deleting via the Sheets API requires a batchUpdate request.
     */
    public function deleteRow(string $spreadsheetId, int $sheetId, int $rowIndex): void
    {
        $request = new Request([
            'deleteDimension' => [
                'range' => [
                    'sheetId'    => $sheetId,
                    'dimension'  => 'ROWS',
                    'startIndex' => $rowIndex - 1, // 0-indexed
                    'endIndex'   => $rowIndex,
                ],
            ],
        ]);
        $batchReq = new BatchUpdateSpreadsheetRequest(['requests' => [$request]]);
        $this->service->spreadsheets->batchUpdate($spreadsheetId, $batchReq);
    }

    /**
     * Look up the numeric sheet ID for a given worksheet title.
     */
    public function getSheetId(string $spreadsheetId, string $worksheetTitle): ?int
    {
        $meta = $this->service->spreadsheets->get($spreadsheetId);
        foreach ($meta->getSheets() as $sheet) {
            if ($sheet->getProperties()->getTitle() === $worksheetTitle) {
                return $sheet->getProperties()->getSheetId();
            }
        }
        return null;
    }
}
